src/cpr_data_access/vespa.py

```diff
@@ -0,0 +1,201 @@
+                (text_block contains ({{stem: false}}"{request.query_string}"))
+            )
+        """
+    else:
+        rendered_query_string_match = f"""
+            where ((
+                {{"targetHits": 1000}} weakAnd(
+                    family_name contains "{ request.query_string }",
+                    family_description contains "{ request.query_string }",
+                    text_block contains "{ request.query_string }"
+                )
+            ) or (
+                [{{"targetNumHits": 1000}}]
+                nearestNeighbor(family_description_embedding,query_embedding)
+            ) or (
+                [{{"targetNumHits": 1000}}]
+                nearestNeighbor(text_embedding,query_embedding)
+            ))
+        """
+
+    rendered_filters = ""
+    if request.keyword_filters:
+        filters = []
+        for field_key, values in request.keyword_filters.items():
+            field_name = filter_fields[field_key]
+            values = [values] if not isinstance(values, list) else values
+            for value in values:
+                filters.append(f'({field_name} contains "{sanitize(value)}")')
+        rendered_filters = " and " + " and ".join(filters)
+
+    if request.year_range:
+        start, end = request.year_range
+        if start:
+            rendered_filters += f" and (family_publication_year >= {start})"
+        if end:
+            rendered_filters += f" and (family_publication_year <= {end})"
+
+    rendered_continuation = (
+        f"{{ 'continuations':['{request.continuation_token}'] }}"
+        if request.continuation_token
+        else ""
+    )
+
+    rendered_query = f"""
+        select *
+        from sources family_document, document_passage
+        { rendered_query_string_match }
+        { rendered_filters }
+        limit 0
+        |
+        { rendered_continuation }
+        all(
+            group(family_import_id)
+            max({request.limit})
+            each(
+                max({request.max_hits_per_family})
+                each(output(summary(search_summary))
+            )
+        )
+    )
+    """
+
+    return " ".join(rendered_query.split())
+
+
+def _parse_vespa_response(
+    request: SearchRequestBody,
+    vespa_response: VespaResponse,
+) -> SearchResponse:
+    """
+    Parse a vespa response into a SearchResponse object
+
+    :param SearchRequestBody request: The user's original search request
+    :param VespaResponse vespa_response: The response from the vespa instance
+    :raises ValueError: if the vespa response status code is not 200, indicating an
+        error in the query, or the vespa instance
+    :return SearchResponse: a list of families, with response metadata
+    """
+    if vespa_response.status_code != 200:
+        raise ValueError(
+            f"Vespa response status code was {vespa_response.status_code}, "
+            f"expected 200"
+        )
```

Comment:
again, classifying this as `ValueError` is only helpful if we're going to catch the `ValueError` around the call to this function & turn it into something a bit more useful. no?